### ISSUE TITLE
[RFC] Broaden PHPUnit version support

### ISSUE THREAD

**Issue by sminnee** [silverstripe/silverstripe-framework#9668]
Right now we're running all our tests on PHPUnit 5. If we want PHPUnit support, we either need to allow the use of PHPUnit 9, or we need to fork PHPUnit 5 and fix it to work with PHP 8, which is likely to be relatively straightforward, but requires the establishment and maintenance of a fork as I have with http://github.com/sminnee/phpunit-mock-objects.

It's worth noting that _users_ of silverstripe/framework needn't care about how the framework's tests are run, they only need to know that their own tests would keep running.

Upgrading PHPUnit would look something like this:

 * Add phpunit 7 | 8 | 9 as a dev dependency of framework, meaning that 7.1 - 9 can be tested.
 * Refactor the core tests such that they pass on those 3 majors
 * Split `src/Dev` out into its own dependency, which can be relegated to a require-dev on client projects, although for BC, silverstripe/framework would require it.
 * Make `src/Dev`, notably SapphireTest, work with phpunit 5 | 6 | 7 | 8 | 9. The advantage of splitting it out as its own dependency is that we can set up a travis config to test against each of these (or at least 5 | 7 | 8 | 9)

Although it's probably more work than forking phpunit 5, it gives us the benefit of being able to use newer versions in our projects. This work also reduces the "everything must be upgraded in lockstep" nature of the project which leads to major releases being so invasive. So it might be worth the work, or at least worth trialling.

The biggest challenge is probably the class rename. Because of this we probably need to have a clause like this:

```
use PHPUnit\Framework\TestCase;
use PHPUnit_Framework_TestCase;

if (!class_exists(TestCase::class)) {
    if (!class_exists(PHPUnit_Framework_TestCase::class)) {
        return;
    };
 
   abstract class TestCase extends PHPUnit_Framework_TestCase {}
}

class SapphireTest extends TestCase
{
}
```

My suggestion is that if we're open to this idea in principle, we start by splitting out Dev (Maybe `silverstripe/dev-support`?) as its own package so that we can start work on a PR that broadens its phpunit requirement and its associated travis matrix.

But, I'm keen to get other feedback.

**Comment by sminnee** [comment]
Having written all this down, I'm now thinking "maybe for starters I'll just publish a fork of phpunit 5.7" :P

**Comment by sminnee** [comment]
OK that's published with a single version - 5.7.28 - http://github.com/sminnee/phpunit and https://packagist.org/packages/sminnee/phpunit

#9665 updates the require-dev requirements to pull this in, among other things.

**Comment by sminnee** [comment]
OK having thought about this more I think that supporting PHPunit 8/9 would be better done with SilverStripe/dev-support:^2, but that such a version would be designed to work with framework 4

**Comment by Acen** [comment]
Was looking at making an RFC around this too. 

Were there any specific reasons around why the version was locked off to 5.7 initially? Considering support dropped off from phpunit themselves at the beginning on 2018.

**Comment by sminnee** [comment]
You can't do a major bump without breaking people's tests. So current projects need to be able to keep using 5.7 if they wish.

But if we altered SapphireTest to support other versions, then we would drop support for phpunit 5.7

Decoupling SapphireTest et al from the rest of framework in dev-support package means that we could use dev-support 2 for testing the framework itself but let existing projects keep using dev-support 1.

**Comment by robbieaverill** [comment]
We also need phpunit 5.x because it supports PHP 5.6, which we still support

**Comment by sminnee** [comment]
Nah we're supporting 7.1+ these days, so we could get away with using phpunit 7 for the core suite and that of supported modules, if we solved the underlying issue.

**Comment by robbieaverill** [comment]
Oh that helps a lot!

**Comment by sminnee** [comment]
OK I've put a bit more detail on what such a module-split might look like.

**Comment by sminnee** [comment]
From @dhensby on the PHP 8 support PR
> It's a shame we can't just upgrade to PHPUnit 8, but there has been some work to the ss-upgrader tool to add an easier upgrade path for that in SS-5.

This RFC is what I've got in mind for that.

**Comment by sminnee** [comment]
One thing I want to highlight here:

Having start on the module-splitting work I think SapphireTest and the other direct PHPUnit bindings are the only that should go into a separate module. The rest should be left in framework, or in the decoupled modules I've been PoCing recently.

So I think the module should be called silverstripe/phpunit-support.

Also: most of the "meat" of SapphireTest should be shifted to more teststate implementations.

This would make it easier to get the benefits of the test system with other test frameworks too, as well as newer phpunit versions.

I'm advocating leaving teststate in core as ORM, Control, Security, etc will have their own state contributions. Pulling these pieces out into a dev module (or multiple dev modules such as dev-orm) will increase coupling.

**Comment by sminnee** [comment]
Giving this more thought, the best first step on this task is probably to:
 * Refactor the guts of SapphireTest into the TestState system
 * Trial the developer experience of project running PHPUnit 9 and avoiding the use of SapphireTest entirely

If it works well it might be easier to simply deprecate the use of SapphireTest altogether, rather than have a module that provides SapphireTest. This would also make it easier to use other test frameworks on projects, such as Codeception.

For modules that are part of recipe-core, recipe-cms, and cwp-recipe-kitchen-sink, we still assumed that the same test tooling is used for all modules. We may want to try decoupling that, but it will be hard.

**Comment by Cheddam** [comment]
A quick update on this after discussion with @sminnee:

We've run into [additional issues](https://github.com/sminnee/phpunit-mock-objects/pull/8#issuecomment-761968412) supporting PHP 8 on the `sminnee/phpunit` forks, and are unlikely to perform any further maintenance on them, as getting the CI green again will require forking further pieces of the `phpunit` ecosystem, and this is an unsustainable and ultimately unhelpful rabbit hole to continue traversing. 

This means that PHP 8 support in 4.x will continue to require use of the `--ignore-platform-reqs` flag for the time being, and any further edgecases or bugs encountered in the `sminnee/phpunit` forks are unlikely to receive fixes.

It's becoming increasingly important to ratify this RFC and proceed with development of PHPUnit 7 / 8 / 9 support in the 4.x line, in order to enable running PHP 8 with no compromises.

**Comment by emteknetnz** [comment]
I'll have a ramble about some problems I see. Apologies in advance, I don't have any solutions here.

**PHP8 support is desirable, but migrating tests seems hard.**
I do recognise the need go on to new versions of phpunit, in particular ^9.3 which includes php8 support.  This is probably the most signficant hurdle to overcome before we can say that Silverstripe has full PHP8 support which means not using the --ignore-platform-reqs flag.

However, the biggest hurdle seems to be migration all the existing silverstripe modules from SapphireTest to a new test system.  You know, all the modules with hundreds of *Test.php files that `use SilverStripe\Dev\SapphireTest; ... class SomeClassTest extends SapphireTest`, well they all need to be updated now.

You cannot composer install two different versions of phpunit i.e. install 5.7 AND 9.3.  This would be pretty useful from a migration point of view so that you can run all the silverstripe unit tests in the vendor folder whether they're written in the new ^9.3 format, or in the old ^5.7 SapphireTest format.  But you can't do this, so unless you could magically update ALL modules at once, a bunch of your unit tests won't be runnable depending on the version on phpunit you have installed.

Also, will things even be installable/dev-buildable if you have phpunit 9 installed because it seems like SapphireTest won't actually exist?

SapphireTest.php
```php
if (!class_exists(PHPUnit_Framework_TestCase::class)) {
    return;
}
class SapphireTest extends PHPUnit_Framework_TestCase implements TestOnly
``

# Forking and technical debt
The easier short-term solution that require no migration is just fork all the relevant phpunit ^5.7 dependencies so they work with PHP8.  This seems like signficantly less work and requires no migration of tests for silverstripe modules.

However there's no denying that we're just taking out further technical debt here.  At some point PHP 7 will be just like PHP 5 and we'll be forced to upgrade.

# Deprecation notices
I don't see deprecating SapphireTest as achieving a whole lot. The module resides in framework so it would only be removed with the release of Silverstripe 5, which as we know is some ways off.  It seems like all this will do in practical terms is spit a bunch of annoying deprecation notices when running unit tests.

**Comment by sminnee** [comment]
One thing about this ticket is that it's not just, or even primarily, about the framework's test themselves – one of the big benefits of this proposal is that it lets _projects_ use newer / alternative testing frameworks, which I think would be useful.

The issues you raise about the core basically come down to any kind of change management, and are separate: "_can_ I use PHPUnit 7||8||9" is different from "should I migrate these 90 module to use it". This ticket is really focused on the first one.

I agree that "deprecation" might be a "little-d deprecation" in that we have better recommended ways for people to write tests (and e.g. we update our docs to avoid discussion of SapphireTest). Formal deprecation should only come after the core and a good set of modules stopped using it.

I'll shift the discussion of PHP8 support to your separate ticket.
